nixos-module.nix

```diff
@@ -1,55 +1,118 @@
-
-let
+{libUtils}: {
+  config,
+  lib,
+  pkgs,
+  ...
+} @ attrs:
+with lib; let
   cfg = config.virtualisation.quadlet;
   quadletUtils = import ./utils.nix {
     inherit lib;
-    systemdLib = (libUtils {
-      inherit lib config pkgs;
-    }).systemdUtils.lib;
+    systemdLib =
+      (libUtils {
+        inherit lib config pkgs;
+      })
+      .systemdUtils
+      .lib;
   };
   # TODO: replace with lib.mergeAttrsList once stable.
   mergeAttrsList = foldl mergeAttrs {};
 
-  containerOpts = types.submodule (import ./container.nix { inherit quadletUtils; } );
-  networkOpts = types.submodule (import ./network.nix { inherit quadletUtils pkgs; } );
+  containerOpts = types.submodule (import ./container.nix {inherit quadletUtils;});
+  networkOpts = types.submodule (import ./network.nix {inherit quadletUtils pkgs;});
+  podOpts = types.submodule (import ./pod.nix {inherit quadletUtils;});
 in {
   options = {
     virtualisation.quadlet = {
       containers = mkOption {
         type = types.attrsOf containerOpts;
-        default = { };
+        default = {};
       };
 
       networks = mkOption {
         type = types.attrsOf networkOpts;
-        default = { };
+        default = {};
+      };
+
+      pods = mkOption {
+        type = types.attrsOf podOpts;
+        default = {};
       };
     };
   };
 
   config = let
-    allObjects = (attrValues cfg.containers) ++ (attrValues cfg.networks);
+    allObjects = (attrValues cfg.containers) ++ (attrValues cfg.networks) ++ (attrValues cfg.pods);
   in {
     virtualisation.podman.enable = true;
+    assertions = let
+      count_occurances = str_list:
+        lib.lists.foldl' (acc: el:
+          if acc ? ${el}
+          then acc // {${el} = acc.${el} + 1;}
+          else acc // {${el} = 1;}) {}
+        str_list;
+      find_duplicate_elements = str_l:
+        lib.attrsets.attrNames
+        (lib.attrsets.filterAttrs (_: v: v > 1)
+          (count_occurances str_l));
+      # assuming that only `name` defines the final name!
+      duplicate_elements = find_duplicate_elements (
+        map
+        (x: x._name)
+        allObjects
+      );
+      missing_pod_suffixes = lib.lists.foldl' (acc: x:
+        if x.containerConfig ? pod && !(lib.strings.hasSuffix ".pod" x.containerConfig.pod)
+        then acc ++ [x._name]
+        else []) [] (attrValues cfg.containers);
+      pods = map (x: x._name) (attrValues cfg.pods);
+      missing_pods = lib.lists.foldl' (acc: x:
+        if x.containerConfig ? pod && !(elem (lib.strings.removeSuffix ".pod" x.containerConfig.pod) pods)
+        then acc ++ [x._name]
+        else []) [] (attrValues cfg.containers);
+    in [
+      {
+        assertion = duplicate_elements == [];
+        message = ''
+          The container/network/pod names should be unique!
+          The following names are not unique: ${lib.strings.concatStringsSep " " duplicate_elements}
+        '';
+      }
+      {
+        assertion = missing_pod_suffixes == [];
```

Comment:
> If we want to be defensive here I think the better way would be to support and encourage cross referencing object names through cfg, which is probably a separate change though.

Could expand on this?

For me, this feature has already helped me a few times, since I always forgot to add `.pod` suffix, so I would like to keep it in some form or another. Though I also understand the benefit of supporting pre-existing pods with names.

Or I could create an option `strict = true` that acts as a filter for the checks? But then I would again be inclined to set _strict_ by default... Or maybe define a global default strict value?

Reply:
I think a better alternative to these checks would be we support and encourage:

```nix
{ config, ... }:
let
    quadletPods = config.virtualisation.quadlet.pods;
{
    virtualisation.quadlet = {
        containers.nginx.containerConfig = {
            pod = quadletPods.nginx-pod.name;  # rather than "nginx-pod.pod"
            # ...
        };
        pods.nginx-pod = {};
    };
};
```

So checks are done by nix language in a way users are familiar with, without we writing and maintaining case-specific checks, while allowing use of external entities wherever needed.

Reply:
Ahh, of course. :man_facepalming: 
Yes, that is definitely better.
For now, I have directly used the interneals `_configName` for the check but it should be now more like the above.

I am hesitant to call it `name` since we have a different `name` option that does not include the suffix in the options above.
Here, I would use whatever you prefer or you can customize it after merging/directly edit the PR.

Reply:
Hmmm, don't like exposing internals.

I'd say let's just get this pod feature in without cross referencing or naming conflict checks (i.e. string literal references for now). I'll finalize those parts afterwards. Everything else LGTM.

Reply:
I've updated the README, and the check has already been removed.
The only check that remains is the one that ensures that a container and pod cannot have the same prefix and links to the docs for further information.

If you would like to apply minor changes, feel free to merge and update it to your liking if this is less work/quicker :)